bootloader/src/pyi_utils.c

```diff
@@ -278,8 +283,53 @@ pyi_get_temp_path(char *buffer, char *runtime_tmpdir)
       /*
        * Set TMP to runtime_tmpdir for _wtempnam() later
        */
-      pyi_path_fullpath(runtime_tmpdir_abspath, PATH_MAX, runtime_tmpdir);
-      pyi_setenv("TMP", runtime_tmpdir_abspath);
+      // Expand environment variables like %LOCALAPPDATA%
+      // We use nested ifs instead of returning immediately on error conditions so that we can fallback
+      // to the default value of %TMP% if we can't process the developer-defined runtime-tmpdir
+      wruntime_tmpdir = pyi_win32_utils_from_utf8(NULL, runtime_tmpdir, 0);
+      if (wruntime_tmpdir) {
+        rc = ExpandEnvironmentStringsW(wruntime_tmpdir, wruntime_tmpdir_expanded, PATH_MAX);
+        free(wruntime_tmpdir);
+        if (rc) {
+          // Get the absolute path
+          wruntime_tmpdir_abspath = _wfullpath(NULL, wruntime_tmpdir_expanded, PATH_MAX);
+          if (wruntime_tmpdir_abspath) {
+            VS("LOADER: absolute runtime tmpdir is %ls\n", wruntime_tmpdir_abspath);
+            // Create the directory path if it does not yet already exist (e.g. %AppData%\NewFolder\NestedFolder)
+            ZeroMemory(path_builder, PATH_MAX * sizeof(wchar_t));
+            cursor = wcschr(wruntime_tmpdir_abspath, L'\\');
+            while(cursor != NULL) {
+              wcsncpy(path_builder, wruntime_tmpdir_abspath, cursor - wruntime_tmpdir_abspath + 1);
+              CreateDirectoryW(path_builder, NULL);
+              // We expect ERROR_ALREADY_EXISTS, ERROR_ACCESS_DENIED (if try to create a drive when running as 
+              // an admin), etc...  If there are true errors that occur when dealing with the specified runtime tmpdir,
+              // those will be elicited below when we try to do mkdir.  Handling errors there will also reset the
+              // TMP environment variable as needed.
+              cursor = wcschr(++cursor, L'\\');
+            }
+            // May not have a string terminated with \, so run CreateDirectoryW one last time to handle that case
+            CreateDirectoryW(wruntime_tmpdir_abspath, NULL);
+            // Store in the TMP environment variable
+            rc = _wputenv_s(L"TMP", wruntime_tmpdir_abspath);
+            free(wruntime_tmpdir_abspath);
+            if (rc) {
+              VS("LOADER: Failed to set the TMP environment variable, falling back to defaults...\n");
+            }
+            else {
+              VS("LOADER: Successfully resolved the specified runtime-tmpdir\n");
+            }
+          }
+          else {
+            VS("LOADER: Failed to obtain the absolute path of the runtime-tmpdir, falling back to defaults...\n");
+          }
+        }
+        else {
+          VS("LOADER: Failed to expand environment variables in the runtime-tmpdir, falling back to defaults...\n");
+        }
+      }
+      else {
+        VS("LOADER: Failed to convert runtime-tmpdir to a wide string, falling back to defaults...\n");
+      }
```

Comment:
If we fail to process the `runtime-tmpdir` option, shouldn't we fail at all (returning 0)?

If so, I suggest to change the code structure into this to avoid deep nesting and keep check and error path close to each other:
```
      wruntime_tmpdir = pyi_win32_utils_from_utf8(NULL, runtime_tmpdir, 0);
      if (! wruntime_tmpdir) {
          VS("LOADER: Failed to convert runtime-tmpdir to a wide string, falling back to defaults...\n");
          return 0 
      }
      rc = ExpandEnvironmentStringsW(wruntime_tmpdir, wruntime_tmpdir_expanded, PATH_MAX);
      if (! rc) {
         …
```